book/src/engine/logic.md

```diff
@@ -0,0 +1,123 @@
+# Logic
+
+## Overview
+
+`chalk-engine` solves a `Goal` using a depth-first search. When asked to solve a
```

Comment:
This isn't strictly true. We don't use depth-first search. If we did, we would always **fully explore** one strand before starting on another. The [docs I wrote before](https://rust-lang.github.io/rustc-guide/traits/slg.html) claimed we use a **breadth-first search** but this isn't really true either. I'm actually not sure if our current search algorithm can be cleanly characterized. It's one of the things I've been wondering about -- we might *want* to think about how to specify the behavior more rigorously. The kind of "disconnected" way that our current solver works makes it a bit less obvious to me how to do that.

The minikanren algorithm is quite interesting in that (by the nature of its design) winds up spending a lot more time traversing and working with answers that have "gotten farther" in the process than it does with those that haven't. Or at least so it was explained to me. To be honest, I have to kind of review how it works, and maybe it'd be good for us to spend some time trying to characterize and work through how our own solver would behave.

Anyway for now let's just not say *depth-first* search, since to me that suggests something very specific (prolog-style solving) which is not the case. Maybe something like "using a hybrid search strategy with elements of depth- and breadth-first search"?

Reply:
Yeah, true, it's not depth- or breadth-first. It would be good to characterize exactly what we do, but I think that will be more useful when we have benches and can check how different search schemes perform.